common/src/Model/GroupNode.cpp

```diff
@@ -35,59 +35,146 @@
 #include "Model/WorldNode.h"
 
 #include <kdl/overload.h>
+#include <kdl/parallel.h>
 #include <kdl/result.h>
 #include <kdl/result_for_each.h>
 #include <kdl/string_utils.h>
 #include <kdl/vector_utils.h>
 
 #include <vecmath/ray.h>
 
+#include <functional>
 #include <string>
+#include <unordered_map>
 #include <vector>
 
 namespace TrenchBroom {
     namespace Model {
+        /**
+         * Given a node, clones its children recursively and applies the given transform.
+         * 
+         * Returns a vector of the cloned direct children of `node`.
+         */
         static kdl::result<std::vector<std::unique_ptr<Node>>, UpdateLinkedGroupsError> cloneAndTransformChildren(const Node& node, const vm::bbox3& worldBounds, const vm::mat4x4& transformation) {
-            using VisitResult = kdl::result<std::unique_ptr<Node>, UpdateLinkedGroupsError>;
-            return kdl::for_each_result(node.children(), [&](const auto* childNode) {
-                return childNode->accept(kdl::overload(
-                    [] (const WorldNode*) -> VisitResult { ensure(false, "Linked group structure is valid"); },
-                    [] (const LayerNode*) -> VisitResult { ensure(false, "Linked group structure is valid"); },
-                    [&](const GroupNode* groupNode) -> VisitResult {
+            // Recursively collect the nodes to clone + transform, starting with the children of `node`.
+            // (`node` itself is skipped.)
+            auto nodesToClone = std::vector<const Model::Node*>{};
+            {
+                std::function<void(const Model::Node*)> collectNodes = [&](const Model::Node* n) {
```

Comment:
Why not use `auto` here? The signature is right next to the declaration.

Reply:
Again, necessary because it's recursive.